integration/testrunner/db.go

```diff
@@ -16,45 +16,68 @@ package testrunner
 
 import (
 	"context"
+	"encoding/json"
+	"errors"
 	"fmt"
 	"testing"
+	"time"
 
 	"cloud.google.com/go/bigquery"
 	"github.com/sethvargo/go-retry"
+	"google.golang.org/api/iterator"
+	"google.golang.org/genproto/googleapis/cloud/audit"
+	"google.golang.org/protobuf/encoding/protojson"
+
+	logpb "cloud.google.com/go/logging/apiv2/loggingpb"
 )
 
-// queryIfAuditLogExists queries the DB and checks if audit log contained in the query exists or not.
-func queryIfAuditLogExists(ctx context.Context, tb testing.TB, query *bigquery.Query, expectedNum int64) (bool, error) {
+// queryAuditLogs queries the DB and checks if audit log contained in the query exists or not and return the results.
+func queryAuditLogs(ctx context.Context, tb testing.TB, query *bigquery.Query) ([]*logpb.LogEntry, error) {
 	tb.Helper()
-
+	time.Sleep(25 * time.Second)
```

Comment:
I find `time.Sleep` calls in tests to _always_ be a smell. Sleeping for 25s seems arbitrary and there's no comment to justify the duration. We should realistically move retry logic and backoff higher in the stack.

Reply:
I have tried that, and the conclusion I made are:
1. If we remove the retry method, there are some other issues that could happen, that we need the retry logic to handle those. 
2. If we remove `time.sleep`, it could be that some logs haven't been logged yet, so we only got partial result, which could make the test fail.

In our tests here, I would have at most 4 logs, so 25 seconds is more than enough.

Reply:
What are the other issues that can happen with retries?

Reply:
I mean there could be problems like "failed to run query", "failed to read job" these kind of issues. With retry, these kind of error can be handled.

Reply:
Sorry. I'm suggesting that we keep the retry and control the logic about when we retry (retryable error vs fatal error). I'm saying that waiting 25s before a test runs seems arbitrary and I would rather us use retry logic instead.

Reply:
Yeah, I have considered that. My concern here is this.

If we want to get 5 logs, and we use retry to keep retrying until we have 5 logs, then this could lead to two issues:
1. This means we need to check the expected log numbers here, which I think should be better handled in the test parts.
2. If the system has error, there are never going to be 5 logs, and we will need to keep retrying until timeout, and in this case, `tb` won't even have the chance to print out the log, when will make it very hard to debug (which has happened before).

And if we don't verify if there are 5 logs in the retry function without time.Sleep(),  retry will quit as soon as some logs are retrieved. But for most cases,  there aren't enough time for logs to be emitted into bigquery, and we won't get enough logs, and the test would fail.

Reply:
Above are my thoughts on why keep using `time.sleep` can be a good option.

Reply:
Sorry I'm still a little confused. We have this function and we have `queryAuditLogsWithRetries` which implements retry logic. If we wait 25s between retries, these tests could quickly take 10m+. Per [go/abcxyz-cicd](http://go/abcxyz-cicd), we aim to have all tests complete in 5m or less.

> If the system has error, there are never going to be 5 logs, and we will need to keep retrying until timeout, and in this case, tb won't even have the chance to print out the log, when will make it very hard to debug (which has happened before).

I agree, but if we retry 4 times, that's 2 minutes of just sleeping and waiting. I don't think this function should have a sleep - the retry function already controls the delay and backoff (L99, L113). It's currently exponential (we should probably switch to fibonacci).

> And if we don't verify if there are 5 logs in the retry function without time.Sleep(), retry will quit as soon as some logs are retrieved. But for most cases, there aren't enough time for logs to be emitted into bigquery, and we won't get enough logs, and the test would fail.

That shouldn't be the case. In the main function, we can check the number of logs and `return retry.RetryableError("not enough logs (expected %d, got %d)")` to force the retry to run again.